org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/folding/FoldingRangeService.java

```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+ * Copyright (c) 2021 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.folding;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.lsp4j.FoldingRange;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.xtext.ide.editor.folding.IFoldingRangeProvider;
+import org.eclipse.xtext.ide.server.Document;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * @author Mark Sujew - Initial contribution and API
+ */
+@Singleton
+public class FoldingRangeService {
+
+	@Inject
+	private IFoldingRangeProvider foldingRangeProvider;
+
+	public List<FoldingRange> createFoldingRanges(Document document, XtextResource resource,
+			CancelIndicator cancelIndicator) {
+		return foldingRangeProvider.getFoldingRanges(resource, cancelIndicator).stream()
+				.map(range -> toFoldingRange(document, range))
+				.filter(range -> range.getStartLine() < range.getEndLine()).collect(Collectors.toList());
+	}
+
+	protected FoldingRange toFoldingRange(Document document, org.eclipse.xtext.ide.editor.folding.FoldingRange range) {
+		int offset = range.getOffset();
+		int length = range.getLength();
+		int endOffset = offset + length;
+		Position start = document.getPosition(offset);
+		Position end = document.getPosition(endOffset);
```

Comment:
Since the endOffset is exclusive, it may be on the next line in the document. I think this needs refinements as the endline is by default fully folded (e.g. if endCharacter is omitted). Do we need to provide the last folded character instead?

Reply:
@Darkgaja Do you have a response to this question?

Reply:
While endOffset is exclusive, if I've understood [this code](https://github.com/eclipse/xtext-core/blob/2b25ff301e6730277b2f270baf29bed2ac490445/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.java#L73-L84) correctly, it shouldn't be a problem. Since the endOffset is then on the newline character, we should still receive the correct line. I am wrong there?

Reply:
Sounds correct to me. How does the folding look like in VS Code e.g. if there is a meaningful char after a closed ML_COMMENT?

```
/*
 * fold me
 */identifier
```

Reply:
The `identifier` part gets folded as well:
![grafik](https://user-images.githubusercontent.com/4377073/108637778-6a253980-748c-11eb-9cd9-c4d8305970fc.png)

Reply:
Interesting. What happens if we also send an end char?

Reply:
Good question. I'd imagine the `identifier` part would still be visible. However, I currently don't have a test setup available where I could easily test this. Can we postpone this question for #1675? (which I currently don't have the time for)

Reply:
Sure